design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,98 @@
+Contributors: TBD
+State: Draft
+
+## Summary
+Support the ability to define how many network devices a VM may potentially contain.
+Allowing the use of hotplug to add network interfaces up to that capacity limit.
+
+## Motivation
+The introduction of network devices hotplug introduced a limitation
+to the amount of possible devices one can attach.
+
+Users of libvirt who can estimate how many potential network devices are needed,
+can define the additional controllers in advance.
+KubeVirt does not expose such domain details through its API.
+But even if these details had been exposed by the VMI object,
+a common user would find it challenging to define such low domain details.
+
+### Goals
+- Allow users to define how many network devices can be potentially hotplug.
+- Do not expose domain architecture details (e.g. PCI/PCIe controllers).
+
+## Proposal
+Users are to express the potential amount of network devices through
+the VMI spec.
+Internally, kubevirt will use libvirt logic to generate the needed
+devices to support the amount specified (e.g. PCIe controllers).
+
+Using hotplug, users may add network devices up to the amount specified
+by the new user *interface-capacity* input.
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- TBD
+
+### API Extensions
+The VMI Spec will be extended to include a new field, named `interfaceCapacity`.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 1 and 32.
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+#### VMI Spec example
+```yaml
+
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-capacity-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+      interfaceCapacity: 5
+      rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the interface-capacity and the current defined interfaces, generate
+  interface-placeholders and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
```

Comment:
Why don't you just add the controllers directly ?

Reply:
I am not interested to know how many and which controllers are needed, nor how to configure them in the domain.
I am looking to defer all this logic to libvirt, which already does this.

Reply:
I guess I'll wait for the example to figure out whatever this means.

Reply:
I have also added it to an Alternative appendix.
Done.